presto-main/src/main/java/com/facebook/presto/operator/aggregation/estimatendv/SingleNDVEstimatorState.java

```diff
@@ -13,11 +13,78 @@
  */
 package com.facebook.presto.operator.aggregation.estimatendv;
 
+import com.facebook.presto.common.array.LongBigArray;
+import com.facebook.presto.common.type.Type;
+import com.facebook.presto.operator.aggregation.histogram.SingleHistogramState;
+import com.facebook.presto.operator.aggregation.histogram.SingleTypedHistogram;
+
+import java.util.HashMap;
+import java.util.Map;
+import java.util.stream.LongStream;
+
 public class SingleNDVEstimatorState
-        extends NDVEstimatorState
+        extends SingleHistogramState
+        implements NDVEstimatorState
 {
-    public SingleNDVEstimatorState()
+    public SingleNDVEstimatorState(Type keyType, int expectedEntriesCount)
+    {
+        super(keyType, expectedEntriesCount);
+    }
+
+    private static Map<Long, Long> computeFreqDict(SingleTypedHistogram histogram)
+    {
+        LongBigArray counts = histogram.getCounts();
+        Map<Long, Long> freqDict = new HashMap<>();
+        int ndv = histogram.getDistinctValueCount();
+        for (int i = 0; i < ndv; i++) {
+            Long count = counts.get(i);
+            freqDict.put(count, freqDict.getOrDefault(count, 0L) + 1);
+        }
+        return freqDict;
+    }
+
+    private long computeBirthdayProblemProbability(long d)
+    {
+        long i = d;
+        long lowerBound;
+        while (true) {
```

Comment:
Is there any chance that this never converges? Is it necessary to add loop count to break out of the loop after some `N` iterations? Also, a javadoc comment referencing the algorithm/paper would be useful.

Reply:
It is guaranteed to converge: the inside function is a monotonically increasing function of `i`, `i` increases during each iteration and it will finally reach a value larger than 10%. 
Javadoc added.